pkg/secrets/vault.go

```diff
@@ -2,23 +2,242 @@ package secrets
 
 import (
 	"fmt"
+	"github.com/hashicorp/vault/api"
+	"io/ioutil"
+	"os"
+	"strings"
 )
 
+type VaultConfig struct {
+	Enabled    bool   `json:"enabled" yaml:"enabled"`
+	Url        string `json:"url" yaml:"url"`
+	AuthMethod string `json:"authMethod" yaml:"authMethod"`
+	Role       string `json:"role" yaml:"role"`
+	Path       string `json:"path" yaml:"path"`
+	Token      string
+}
+
 type VaultSecret struct {
 	engine        string
-	namespace     string
+	path          string
 	key           string
-	base64Encoded bool
+	base64Encoded string
 }
 
 type VaultDecrypter struct {
-	secretConfig string
+	encryptedSecret string
+	vaultConfig     VaultConfig
+	isKV2           map[string]struct{}
+}
+
+func NewVaultDecrypter(encryptedSecret string) *VaultDecrypter {
+	return &VaultDecrypter{encryptedSecret, VaultConfig{}, map[string]struct{}{}}
 }
 
-func NewVaultDecrypter(secretConfig string) *VaultDecrypter {
-	return &VaultDecrypter{secretConfig}
+func getVaultDecrypter(encryptedSecret string) Decrypter {
+	decrypter := NewVaultDecrypter(encryptedSecret)
+	decrypter.vaultConfig = Registry.VaultConfig
+	return decrypter
 }
 
 func (v *VaultDecrypter) Decrypt() (string, error) {
-	return "", fmt.Errorf("vault secrets unsupported")
+	if err := v.ValidateVaultConfig(); err != nil {
+		return "", fmt.Errorf("vault configuration error - %s", err)
+	}
+
+	vaultSecret, err := ParseVaultEncryptedSecret(v.encryptedSecret)
+	if err != nil {
+		return "", fmt.Errorf("error parsing vault secret syntax - %s", err)
+	}
+
+	if v.vaultConfig.Token == "" {
+		token, err := v.FetchVaultToken()
+		if err != nil {
+			return "", fmt.Errorf("error fetching vault token - %s", err)
+		}
+		v.vaultConfig.Token = token
+	}
+
+	secret, err := v.FetchSecret(vaultSecret)
+	if err != nil {
```

Comment:
should we check the type of error before retrying? It should be not found, not 403 etc..

Reply:
Updated, but not sure how to best check this error...